File: sdk/js/src/models/ChatCompletionResult.js
/**
 * ChatCompletionResult class representing the result of a non-streaming chat completion.
 */
export default class ChatCompletionResult {
  /**
   * @param {Object} result - Information about the completion result.
   * @param {string} [result.ThreadGUID] - Globally unique identifier for the thread.
   * @param {string} [result.TurnGUID] - Globally unique identifier for the turn.
   * @param {string|null} [result.Message=null] - The assistant's message (default is null).
   * @param {string|null} [result.Reasoning=null] - Reasoning/thinking content, if any (default is null).
   * @param {string} [result.Provider='OpenAI'] - Provider type used for the completion (default is OpenAI).
   * @param {string|null} [result.Model=null] - Model name used (default is null).
   * @param {number|null} [result.PromptTokens=null] - Prompt token count (default is null).
   * @param {number|null} [result.CompletionTokens=null] - Completion token count (default is null).
   * @param {number|null} [result.TimeToFirstTokenMs=null] - Time to first token in milliseconds (default is null).
   * @param {number|null} [result.TimeToLastTokenMs=null] - Time to last token in milliseconds (default is null).
   * @param {number} [result.TotalDurationMs=0] - Total duration in milliseconds (default is 0).
   * @param {number|null} [result.TokensPerSecondOverall=null] - Overall tokens per second (default is null).
   * @param {number} [result.ToolCallCount=0] - Number of tool calls (default is 0).
   * @param {number} [result.ToolLoopIterations=0] - Number of tool loop iterations (default is 0).
   * @param {number} [result.RetrievedChunkCount=0] - Number of retrieved RAG chunks (default is 0).
   * @param {number} [result.RetryCount=0] - Number of retries performed (default is 0).
   */
  constructor(result = {}) {
    const {
      ThreadGUID = null,
      TurnGUID = null,
      Message = null,
      Reasoning = null,
      Provider = 'OpenAI',
      Model = null,
      PromptTokens = null,
      CompletionTokens = null,
      TimeToFirstTokenMs = null,
      TimeToLastTokenMs = null,
      TotalDurationMs = 0,
      TokensPerSecondOverall = null,
      ToolCallCount = 0,
      ToolLoopIterations = 0,
      RetrievedChunkCount = 0,
      RetryCount = 0,
    } = result;

    this.ThreadGUID = ThreadGUID; // Unique identifier for the thread
    this.TurnGUID = TurnGUID; // Unique identifier for the turn
    this.Message = Message; // Assistant message
    this.Reasoning = Reasoning; // Reasoning content
    this.Provider = Provider; // Provider type
    this.Model = Model; // Model name
    this.PromptTokens = PromptTokens; // Prompt token count
    this.CompletionTokens = CompletionTokens; // Completion token count
    this.TimeToFirstTokenMs = TimeToFirstTokenMs; // Time to first token
    this.TimeToLastTokenMs = TimeToLastTokenMs; // Time to last token
    this.TotalDurationMs = TotalDurationMs; // Total duration
    this.TokensPerSecondOverall = TokensPerSecondOverall; // Overall tokens per second
    this.ToolCallCount = ToolCallCount; // Tool call count
    this.ToolLoopIterations = ToolLoopIterations; // Tool loop iterations
    this.RetrievedChunkCount = RetrievedChunkCount; // Retrieved chunk count
    this.RetryCount = RetryCount; // Retry count
  }
}

File: sdk/csharp/src/LiteGraph.Sdk/ChatCompletionResult.cs
namespace LiteGraph.Sdk
{
    using System;

    /// <summary>
    /// Chat completion result, returned as the non-streaming response body and inside the streaming usage event.
    /// </summary>
    public class ChatCompletionResult
    {
        #region Public-Members

        /// <summary>
        /// Thread GUID.
        /// </summary>
        public Guid ThreadGUID { get; set; } = Guid.NewGuid();

        /// <summary>
        /// Turn GUID.
        /// </summary>
        public Guid TurnGUID { get; set; } = Guid.NewGuid();

        /// <summary>
        /// Assistant message.
        /// </summary>
        public string Message { get; set; } = null;

        /// <summary>
        /// Model reasoning text.  Null when the model emitted none.
        /// </summary>
        public string Reasoning { get; set; } = null;

        /// <summary>
        /// Provider of the completion endpoint.  Default is OpenAI.
        /// </summary>
        public ChatProviderTypeEnum Provider { get; set; } = ChatProviderTypeEnum.OpenAI;

        /// <summary>
        /// Model that generated the response.
        /// </summary>
        public string Model { get; set; } = null;

        /// <summary>
        /// Prompt tokens reported by the provider.  Null when unreported.
        /// </summary>
        public int? PromptTokens { get; set; } = null;

        /// <summary>
        /// Completion tokens reported by the provider.  Null when unreported.
        /// </summary>
        public int? CompletionTokens { get; set; } = null;

        /// <summary>
        /// Time to first token in milliseconds.  Null when unknown.
        /// </summary>
        public double? TimeToFirstTokenMs { get; set; } = null;

        /// <summary>
        /// Time to last token in milliseconds.  Null when unknown.
        /// </summary>
        public double? TimeToLastTokenMs { get; set; } = null;

        /// <summary>
        /// Total turn duration in milliseconds.
        /// </summary>
        public double TotalDurationMs { get; set; } = 0;

        /// <summary>
        /// Overall tokens per second.  Null when unknown.
        /// </summary>
        public double? TokensPerSecondOverall { get; set; } = null;

        /// <summary>
        /// Number of tool calls executed within the turn.
        /// </summary>
        public int ToolCallCount { get; set; } = 0;

        /// <summary>
        /// Number of tool loop iterations within the turn.
        /// </summary>
        public int ToolLoopIterations { get; set; } = 0;

        /// <summary>
        /// Number of retrieved context chunks injected into the prompt.
        /// </summary>
        public int RetrievedChunkCount { get; set; } = 0;

        /// <summary>
        /// Retries performed before the response started.
        /// </summary>
        public int RetryCount { get; set; } = 0;

        #endregion

        #region Private-Members

        #endregion

        #region Constructors-and-Factories

        /// <summary>
        /// Instantiate.
        /// </summary>
        public ChatCompletionResult()
        {

        }

        #endregion

        #region Public-Methods

        #endregion

        #region Private-Methods

        #endregion
    }
}
